### PR TITLE
Upstream compat

### PR DESCRIPTION
**Pull request by TomAugspurger**
xref https://github.com/dask/dask/issues/6148

Closes https://github.com/dask/dask/issues/6139

### CODE REVIEW

**Comment by mrocklin** [comment]
+1

**Comment by jrbourbeau** [comment]
Just as a heads up, `numba=0.49.0` isn't available on conda-forge yet (xref https://github.com/conda-forge/numba-feedstock/pull/44)

**Comment by TomAugspurger** [comment]
Also "fixing" the numpy tests here.

NumPy as some issues with the truthiness of strings of just whitespace.

```python
In [12]: np.argwhere(["a", " ", "c"])
Out[12]:
array([[0],
       [2]])
```

Confirming the future behavior in https://github.com/numpy/numpy/issues/9875#issuecomment-622457875, but I think we just avoid it for now.

**Comment by jakirkham** [comment]
Thanks for working on this Tom! 😄

When I added `count_nonzero` and `argwhere`, was trying to align with NumPy. Though also opened issues upstream, of which the xref'd issue comes out of. My hope is that as NumPy improves the handling of `str`s in NumPy arrays; we can simplify the logic we have here. Certainly it may be a bit bumpy until then. Happy with whatever you see fit to smooth over this.

**Comment by TomAugspurger** [comment]
Thanks. I'm not trying to match numpy's behavior.

**Comment by TomAugspurger** [comment]
This should be good now. The only remaining upstream failures will be fixed in https://github.com/dask/dask/pull/6156

File: dask/array/routines.py [hunk 1]
```diff
@@ -1188,6 +1188,13 @@ def _isnonzero_vec(v):
 
 
 def isnonzero(a):
+    if a.dtype.kind in {"U", "S"}:
+        # NumPy treats all-whitespace strings as falsy in some places (where).
+        # but not in `.astype(bool)`. To match the behavior of numpy at least until
+        # 1.19, we is _isnonzero_vec. When NumPy changes behavior, we should just
```
Comment (jakirkham) [inline dask/array/routines.py:1194]:
```suggestion
        # 1.19, we use `_isnonzero_vec`. When NumPy changes behavior, we should just
```

File: dask/array/routines.py [hunk 2]
```diff
@@ -1188,6 +1188,13 @@ def _isnonzero_vec(v):
 
 
 def isnonzero(a):
+    if a.dtype.kind in {"U", "S"}:
+        # NumPy treats all-whitespace strings as falsy in some places (where).
```
Comment (jakirkham) [inline dask/array/routines.py:1192]:
```suggestion
        # NumPy treats all-whitespace strings as falsy (like in `np.nonzero`).
```

**Comment by jakirkham** [comment]
Thanks Tom! 😄

**Pull request merged**